File: code/2022-03-06-thinking/add_two.py
"""The little addition program in Python."""

def add_two(a, b):
    return a + b

# this works
assert add_two(1, 2) == 3

class Bar:
    def __add__(self, other):
        return Bar()

a = Bar()
b = Bar()
assert add_two(a, b) == Bar()   # this works

class Foo:
    pass

a = Foo()
b = Foo()
add_two(a, b)   # this won't work

File: code/2022-03-06-thinking/add_two.rs
/// The addition program in Rust.

// this will compile
fn add_two<T: std::ops::Add<Output = T>>(a: T, b: T) -> T {
    a + b
}

// this won't
// fn add_two<T: std::ops::Add>(a: T, b: T) -> T {
//     a + b
// }

// this won't either
// fn add_two<T>(a: T, b: t) -> T {
//     a + b
// }

fn main() {
    assert_eq!(add_two(1, 2), 3);
}
